simba/rotation.py

```diff
@@ -27,52 +27,80 @@ def add_trip(self, trip):
+            if self.departure_time > trip.departure_time:
                 # first of rotation found (for now)
-                self.departure_time = new_trip.departure_time
-                self.departure_name = new_trip.departure_name
+                self.departure_time = trip.departure_time
+                self.departure_name = trip.departure_name
             # '<=' instead of '<' since last trip of rotation has no duration
             # in the sample data. The trips are however chronologically
             # sorted which is why this approach works for sample data.
             # Will also work if one only relies on timestamps!
-            elif self.arrival_time <= new_trip.arrival_time:
+            elif self.arrival_time <= trip.arrival_time:
                 # last trip of rotation (for now)
-                self.arrival_time = new_trip.arrival_time
-                self.arrival_name = new_trip.arrival_name
-
-        # set charging type if given
-        charging_type = trip.get('charging_type')
-        self.trips.append(new_trip)
-        if charging_type in ['depb', 'oppb']:
-            assert self.schedule.vehicle_types.get(
-                self.vehicle_type, {}).get(charging_type) is not None, (
-                f"The required vehicle type {self.vehicle_type}({charging_type}) "
-                "is not given in the vehicle_types.json file.")
-            if self.charging_type is None:
-                # set CT for whole rotation
-                self.set_charging_type(charging_type)
-            elif self.charging_type == charging_type:
-                # same CT as other trips: just add trip consumption
-                self.consumption += new_trip.calculate_consumption()
-            else:
-                # different CT than rotation: error
-                raise Exception(f"Two trips of rotation {self.id} have distinct charging types")
+                self.arrival_time = trip.arrival_time
+                self.arrival_name = trip.arrival_name
+
+        if self.vehicle_type is None:
+            self.vehicle_type = trip.vehicle_type
+            for prior_trip in self.trips:
+                prior_trip.vehicle_type = trip.vehicle_type
+        else:
+            assert self.vehicle_type == trip.vehicle_type, (
+                f"Two trips of rotation {self.id} have distinct vehicle types")
+
+        if self.charging_type is not None and self.charging_type != trip.charging_type:
+            raise ValueError(f"Two trips of rotation {self.id} have distinct charging types")
+
+        # (re)calculate consumption
+        # might change if charging_type changes
+        # therefore, add new trip only after charging_type is set
+        trip.rotation = self
+        if trip.charging_type is None:
+            trip.charging_type = self.charging_type
+        if trip.charging_type in ["oppb", "depb"]:
+            self.set_charging_type(trip.charging_type)
+            self.consumption += trip.calculate_consumption()
+
```

Comment:
if i read this correctly, in cases where the rotation has ct=None, added trips do not add to the consumption. A calculation of all trips with ct=None would be needed at the point where ct is set. In my opinion continuous calculation of consumption should be simply removed from add_trips. 
the pre_simulation already takes care of calculating the consumption when the schedule is generated.
Also I would like it better if the same ct and vt are enforced for a rotation instead of allowing a mix of None and ct/vt

Reply:
removed consumption update from add_trip